server/gateway.go

```diff
@@ -0,0 +1,1707 @@
+		s.gateway.Lock()
+		s.gateway.rqs[string(sub.sid)] = sub
+		s.gateway.Unlock()
+	}
+	var (
+		_gws [4]*client
+		gws  = _gws[:0]
+	)
+	s.getInboundGateways(&gws)
+	if len(gws) == 0 {
+		return
+	}
+	var (
+		protoa = [512]byte{}
+		proto  = protoa[:0]
+	)
+	for _, c := range gws {
+		// Always send for queues. Note that by contract this function
+		// is called only once per account/subject/queue.
+		sendProto := sub.queue != nil
+		// We still want to execute following code even if this is a
+		// queue and we are going to send the proto anyway, because
+		// it may clear the noInterest map if we are going from
+		// a no-interest to a queue sub.
+		if sni, _ := c.gw.noInterest.Load(accName); sni != nil {
+			sn := sni.(*sync.Map)
+			if _, noInterest := sn.Load(string(sub.subject)); noInterest {
+				sn.Delete(string(sub.subject))
+				sendProto = true
+			}
+		}
+		if sendProto {
+			if len(proto) == 0 {
+				proto = append(proto, rSubBytes...)
+				proto = append(proto, accName...)
+				proto = append(proto, ' ')
+				proto = append(proto, sub.subject...)
+				if sub.queue != nil {
+					proto = append(proto, ' ')
+					proto = append(proto, sub.queue...)
+					// For now, just use 1 for the weight
+					proto = append(proto, ' ', '1')
+				}
+				proto = append(proto, CR_LF...)
+			}
+			c.mu.Lock()
+			if c.trace {
+				c.traceOutOp("", proto[:len(proto)-LEN_CR_LF])
+			}
+			c.sendProto(proto, true)
+			c.mu.Unlock()
+		}
+	}
+}
+
+func (s *Server) sendQueueUnsubToGateways(accName string, qsub *subscription) {
+	// If this is a remote queue sub, need to remove from remote queue subs map
+	if qsub.client != nil && qsub.client.typ == ROUTER {
+		s.gateway.Lock()
+		delete(s.gateway.rqs, string(qsub.sid))
+		s.gateway.Unlock()
+	}
+	var (
+		_gws [4]*client
+		gws  = _gws[:0]
+	)
+	s.getInboundGateways(&gws)
+	if len(gws) == 0 {
+		return
+	}
+	var (
+		protoa = [512]byte{}
+		proto  = protoa[:0]
+	)
+	proto = append(proto, rUnsubBytes...)
+	proto = append(proto, accName...)
+	proto = append(proto, ' ')
+	proto = append(proto, qsub.subject...)
+	proto = append(proto, ' ')
+	proto = append(proto, qsub.queue...)
+	proto = append(proto, CR_LF...)
+	for _, c := range gws {
+		c.mu.Lock()
+		if c.trace {
+			c.traceOutOp("", proto[:len(proto)-LEN_CR_LF])
+		}
+		c.sendProto(proto, true)
+		c.mu.Unlock()
+	}
+}
+
+// May send a message to all outbound gateways. It is possible
+// that message is not sent to a given gateway if for instance
+// it is known that this gateway has no interest in account or
+// subject, etc..
+func (c *client) sendMsgToGateways(msg, subject, reply []byte, doQueues bool) {
+	var (
+		gwsa [4]*client
+		gws  = gwsa[:0]
+		s    = c.srv
+	)
+	s.getOutboundGateways(&gws)
+	if len(gws) == 0 {
+		return
+	}
+	var (
+		subj    = string(subject)
+		queuesa = [512]byte{}
+		queues  = queuesa[:0]
+		groups  = map[string]struct{}{}
+		accName = c.acc.Name
+	)
+	if doQueues {
```

Comment:
What if we have multiple queue groups? This seems binary.

Reply:
ok follow on question. If I have qgroup for foo:bar locally in my cluster, but a GW has foo:baz, this will work correctly yes?

Reply:
hum... no. The doQueues would be set to false since len(r.qsubs) in process inbound clients would not be 0. So I guess it is more complex. I need to gather the accounts/queue groups that have been used in local/cluster and pass that to this function...

Reply:
I think so. The GWs have knowledge of who they can send to via interest in queue subs from the other side. We just need to have a local suppression list/map of queue groups we already delivered to.

Reply:
That suppression map is what I was doing in sendMsgToGateways, but it needs to be pre-populated with the queues that were used in processClientInboundMsg.

Reply:
yep. So top of processClientInboundMsg needs to see if there are GWs defined etc and track. Maps you can't reset so might be worth just doing simple list in that func and pass to GWs.

Reply:
I was thinking map on the stack like I did in sendMsgToGateway() and pass it (map are passed by reference)

Reply:
ok they still have a cost to setup but I say start with that.